Library/OcCryptoLib/X64/Sha2-avx.c

```diff
@@ -0,0 +1,227 @@
+/** @file
+  Copyright (C) 2021, ISP RAS. All rights reserved.
+
+  All rights reserved.
+
+  This program and the accompanying materials
+  are licensed and made available under the terms and conditions of the BSD License
+  which accompanies this distribution.  The full text of the license may be found at
+  http://opensource.org/licenses/bsd-license.php
+
+  THE PROGRAM IS DISTRIBUTED UNDER THE BSD LICENSE ON AN "AS IS" BASIS,
+  WITHOUT WARRANTIES OR REPRESENTATIONS OF ANY KIND, EITHER EXPRESS OR IMPLIED.
+**/
+
+#include "../Sha2-avx.h"
+#include <Library/BaseLib.h>
+#include <Library/BaseMemoryLib.h>
+
+//
+// Sha 512 functions
+//
+VOID
+EFIAPI
+Sha512TransformAvx (
+  IN OUT SHA512_STATE  *State,
+  IN     CONST UINT8   *Data,
+  IN     UINTN         BlockNb
+  );
+
+VOID
+Sha512InitAvx (
+  IN OUT SHA512_STATE  *Context
+  )
+{
+  UINTN  Index;
+
+  for (Index = 0; Index < 8; ++Index) {
+    Context->State[Index] = SHA512_H0[Index];
```

Comment:
Cannot we use CopyMem here?

Also, I get a feeling that after the unification you would only have custom versions of Sha512UpdateAvx / Sha384UpdateAvx with Init and Final being common between the implementations.